lib/net.js

```diff
@@ -501,8 +502,15 @@ Socket.prototype.setKeepAlive = function(setting, msecs) {
     return this;
   }
 
-  if (this._handle.setKeepAlive)
-    this._handle.setKeepAlive(setting, ~~(msecs / 1000));
+  const newValue = [setting, ~~(msecs / 1000)];
+  if (this._handle.setKeepAlive &&
+   (this[kSetKeepAlive] == null ||
+    this[kSetKeepAlive][0] !== newValue[0] ||
+    this[kSetKeepAlive][1] !== newValue[1])
+  ) {
+    this[kSetKeepAlive] = newValue;
+    this._handle.setKeepAlive(...newValue);
+  }
```

Comment:
This seems a bit more complicated than it needs to be, and there's another optimization here that can be made.

First, the optimization: in the docs, it points out that when `initialDelay` (`msecs` in the function) is equal to `0`, the value will be left unchanged, so we can likely include a check for that here to avoid the call in that case.

Second, on simplifying, `msecs` is only relevant when `setting` is toggled, so perhaps we don't actually have to cache that.

```js
if (this._handle.setKeepAlive && this[kSetKeepAlive] !== setting) {
  this._handle.setKeepAlive(setting, msecs);
  this[kSetKeepAlive] = setting;
}
```

I would also prefer that this has proper type checking for setting and msecs but doing so would make this semver-major.

For example, a check for `setting` would be:
```js
if (typeof setting !== 'boolean')
  throw new ERR_INVALID_ARG_TYPE('enable', 'boolean', setting);
```

Reply:
@jasnell Thank you for your feedback, I've incorporated your suggestions in my latest commit.